pageserver/src/walingest.rs

```diff
@@ -104,12 +103,66 @@ struct WarnIngestLag {
     timestamp_invalid_msg_ratelimit: RateLimit,
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum WalIngestError {
+    #[error(transparent)]
+    #[allow(private_interfaces)]
+    PageReconstructError(#[from] PageReconstructError),
+    #[error(transparent)]
+    DeserializationFailure(#[from] DeserializeError),
+    #[error(transparent)]
+    SerializationFailure(#[from] SerializeError),
+    #[error("the request contains data not supported by pageserver: {0} @ {1}")]
+    InvalidKey(Key, Lsn),
+    #[error("twophase file for xid {0} already exists")]
+    FileAlreadyExists(u64),
+    #[error("slru segment {0:?}/{1} already exists")]
+    SlruAlreadyExists(SlruKind, u32),
+    #[error("relation already exists")]
+    RelationAlreadyExists,
+    #[error("invalid relnode")]
+    InvalidRelnode,
```

Comment:
I'd use `AssertionError` for these since they're shallow errors without much content.

Reply:
made it use `LogicalError`.

Reply:
I don't really want `WalIngestError` to become a shallow `anyhow::Error` wrapper. Then it becomes too easy for people when they encounter an `anyhow::Error`, maybe from a place that returns a cancellation error, to "just" wrap it into an `LogicalError`, even though `LogicalError`s are supposed to have their inner `anyhow::Error`s to always be constructed locally.

Reply:
I think all the variants in the enum are a good fit for logical error. Your point about future errors being wrapped in `LogicalError` stands anyway (even if we keep this as is, I can still misuse logical error in new code). Can mitigate that with a comment on `LogicalError`: "use this for when you detect an invariant violation on the ingest path".

Reply:
I'm still not convinced. feel free to file a follow-up.